R/PipeOpBLSmote.R

```diff
@@ -0,0 +1,124 @@
+#'
+#' @section Input and Output Channels:
+#' Input and output channels are inherited from [`PipeOpTaskPreproc`].
+#'
+#' The output during training is the input [`Task`][mlr3::Task] with added synthetic rows for the minority class.
+#' The output during prediction is the unchanged input.
+#'
+#' @section State:
+#' The `$state` is a named `list` with the `$state` elements inherited from [`PipeOpTaskPreproc`].
+#'
+#' @section Parameters:
+#' The parameters are the parameters inherited from [`PipeOpTaskPreproc`], as well as:
+#' * `K` :: `numeric(1)` \cr
+#'   The number of nearest neighbors used for sampling from the minority class. Default is `5`.
+#'   See [`BLSMOTE()`][`smotefamily::BLSMOTE`].
+#' * `C` :: `numeric(1)` \cr
+#'   The number of nearest neighbors used for classifying sample points as SAFE/DANGER/NOISE. Default is `5`.
+#'   See [`BLSMOTE()`][`smotefamily::BLSMOTE`].
+#' * `dup_size` :: `numeric` \cr
+#'   Desired times of synthetic minority instances over the original number of majority instances. `0` leads to balancing minority and majority class.
+#'   Default is `0`. See [`BLSMOTE()`][`smotefamily::BLSMOTE`].
+#' * `method` :: `character(1)` \cr
+#'   The type of Borderline-SMOTE algorithm to use. Default is `"type1"`.
+#'   See [`BLSMOTE()`][`smotefamily::BLSMOTE`].
+#'
+#' @section Fields:
+#' Only fields inherited from [`PipeOpTaskPreproc`]/[`PipeOp`].
+#'
+#' @section Methods:
+#' Only methods inherited from [`PipeOpTaskPreproc`]/[`PipeOp`].
+#'
+#' @references
+#' `r format_bib("han_2005")`
+#'
+#' @family PipeOps
+#' @template seealso_pipeopslist
+#' @include PipeOpTaskPreproc.R
+#' @export
+#' @examples
+#' \dontshow{ if (requireNamespace("smotefamily")) \{ }
+#' library("mlr3")
+#'
+#' # Create example task
+#' data = smotefamily::sample_generator(500, 0.8)
+#' data$result = factor(data$result)
+#' task = TaskClassif$new(id = "example", backend = data, target = "result")
+#' task$head()
+#' table(task$data(cols = "result"))
+#'
+#' # Generate synthetic data for minority class
+#' pop = po("blsmote")
+#' bls_result = pop$train(list(task))[[1]]$data()
+#' nrow(bls_result)
+#' table(bls_result$result)
+#' \dontshow{ \} }
+PipeOpBLSmote = R6Class("PipeOpBLSmote",
+  inherit = PipeOpTaskPreproc,
+  public = list(
+    initialize = function(id = "blsmote", param_vals = list()) {
+      ps = ps(
+        K = p_int(lower = 1, default = 5, tags = c("train", "blsmote")),
+        C = p_int(lower = 1, default = 5, tags = c("train", "blsmote")),
+        # dup_size = 0 leads to behaviour different from 1, 2, 3, ..., because it means "duplicating until balanced", so it is a 'special_vals'.
+        dupSize = p_int(lower = 1, default = 0, special_vals = list(0), tags = c("train", "blsmote")),
+        # Default of `method` is derived from the source code of smotefamily::BLSMOTE(), not documented there.
+        method = p_fct(levels = c("type1", "type2"), tags = c("train", "blsmote"))
+      )
```

Comment:
add hyperparameter "quiet", set to TRUE, *on default*, use `capture.output()` to suppress print in train.